x-pack/plugins/security_solution/public/explore/network/components/embeddables/embedded_map.tsx

```diff
@@ -123,47 +123,48 @@ export const EmbeddedMapComponent = ({
   const isFieldInIndexPattern = useIsFieldInIndexPattern();
 
   const [mapDataViews, setMapDataViews] = useState<SourcererDataView[]>([]);
+  const [availableDataViews, setAvailableDataViews] = useState<SourcererDataView[]>([]);
 
-  const availableDataViews = useMemo(() => {
+  const isMountedRef = useRef(true);
+  useEffect(() => {
+    return () => {
+      isMountedRef.current = false;
+    };
+  }, []);
+
+  const fetchData = useCallback(async () => {
+    try {
+      const apiResponse = await Promise.all(
+        availableDataViews.map(async ({ title }) =>
+          isFieldInIndexPattern(title, getRequiredMapsFields(title))
+        )
+      );
+      // ensures only index patterns with maps fields are passed
+      const goodDataViews = availableDataViews.filter((_, i) => apiResponse[i] ?? false);
+      if (isMountedRef.current) {
+        setMapDataViews(goodDataViews);
+      }
+    } catch (e) {
+      if (isMountedRef.current) {
+        setMapDataViews([]);
+        addError(e, { title: i18n.ERROR_CREATING_EMBEDDABLE });
+        setIsError(true);
+      }
+    }
+  }, [addError, availableDataViews, isFieldInIndexPattern]);
+
+  useEffect(() => {
```

Comment:
I had a hard time here, `useMemo` and `useState` are not working as i expect. Why does `availableDataViews` get logged (rerendered) 3 times with `useMemo` and only once (with data) with `useState`/`useEffect`? I don't think I should even have to be doing that `isEqual` check, shouldn't the `setState` function handle that?

This was confusing to me and I still don't understand why react is behaving this way, if anyone can enlighten me... the 2 different versions of the code are below. Why does `availableDataViews` get logged (rerendered) 3 times with `useMemo` and only once (with data) with `useState`/`useEffect`?

## `useMemo` code:
```
  const availableDataViews = useMemo(() => {
    const dataViews = kibanaDataViews.filter((dataView) =>
      selectedPatterns.includes(dataView.title)
    );
    if (selectedPatterns.length > 0 && dataViews.length === 0) {
      setIsIndexError(true);
    }
    return dataViews;
  }, [kibanaDataViews, selectedPatterns]);

  useEffect(() => {
    console.log('mount');
  }, []);

  useEffect(() => {
    console.log('availableDataViews', availableDataViews);
  }, [availableDataViews]);
```
## `useMemo` logs:
```
  // logs
  availableDataViews (2)[{…}, {…}]
  mount
  availableDataViews (2)[{…}, {…}]
  availableDataViews (2)[{…}, {…}]
```

## `useState` code
```
 const [availableDataViews, setAvailableDataViews] = useState<SourcererDataView[]>([]);

   useEffect(() => {
    const dataViews = kibanaDataViews.filter((dataView) =>
      selectedPatterns.includes(dataView.title)
    );
    if (selectedPatterns.length > 0 && dataViews.length === 0) {
      setIsIndexError(true);
    }
    if (!isEqual(availableDataViews, dataViews)) {
      setAvailableDataViews(dataViews);
      fetchData();
    }
  }, [availableDataViews, fetchData, kibanaDataViews, selectedPatterns]);

  useEffect(() => {
    console.log('mount');
  }, []);

  useEffect(() => {
    console.log('availableDataViews', availableDataViews);
  }, [availableDataViews]);
```
## After logs
```
  // logs
  availableDataViews (2)[]
  mount
  availableDataViews (2)[{…}, {…}]
```

Reply:
The technique above saves 2 additional fields api calls per index pattern